File: client/packages/levee-driver/src/msgpackSerializer.ts
/**
 * MessagePack encode/decode functions for Phoenix Socket.
 *
 * @remarks
 * These functions match the `encode` and `decode` option interface expected
 * by the Phoenix JS client's Socket constructor. When used together with
 * `vsn: "3.0.0"`, the server selects the matching MsgpackSerializer.
 *
 * Wire format: a msgpack-encoded array `[join_ref, ref, topic, event, payload]`.
 *
 * @packageDocumentation
 */

import { pack, unpack } from "msgpackr";

/**
 * Shape of a Phoenix channel message used by the Socket's encode/decode hooks.
 */
interface PhoenixMessage {
	join_ref: string | null;
	ref: string | null;
	topic: string;
	event: string;
	payload: unknown;
}

/**
 * Encodes a Phoenix message as a msgpack binary payload.
 *
 * @param msg - The message to encode
 * @param callback - Callback receiving the encoded Uint8Array
 */
export function encodeMsgpack(
	msg: object,
	callback: (encoded: unknown) => void,
): void {
	const m = msg as PhoenixMessage;
	const data = pack([m.join_ref, m.ref, m.topic, m.event, m.payload]);
	callback(data);
}

/**
 * Decodes a msgpack binary payload into a Phoenix message.
 *
 * @param rawPayload - The raw binary data from the WebSocket
 * @param callback - Callback receiving the decoded message
 */
export function decodeMsgpack(
	rawPayload: unknown,
	callback: (decoded: object) => void,
): void {
	const decoded = unpack(new Uint8Array(rawPayload as ArrayBuffer)) as [
		string | null,
		string | null,
		string,
		string,
		unknown,
	];
	callback({
		join_ref: decoded[0],
		ref: decoded[1],
		topic: decoded[2],
		event: decoded[3],
		payload: decoded[4],
	});
}

File: server/lib/levee_web/channels/msgpack_serializer.ex
defmodule LeveeWeb.MsgpackSerializer do
  @moduledoc """
  A Phoenix Socket serializer that uses MessagePack for binary WebSocket frames.

  Selected when a client connects with `vsn=3.0.0`. Messages use the same
  logical 5-element array format as the JSON serializer:

      [join_ref, ref, topic, event, payload]

  but are encoded as msgpack binary instead of JSON text.
  """

  @behaviour Phoenix.Socket.Serializer

  alias Phoenix.Socket.{Broadcast, Message, Reply}

  @impl true
  def fastlane!(%Broadcast{} = msg) do
    data = Msgpax.pack!([nil, nil, msg.topic, msg.event, msg.payload])
    {:socket_push, :binary, data}
  end

  @impl true
  def encode!(%Reply{} = reply) do
    data =
      Msgpax.pack!([
        reply.join_ref,
        reply.ref,
        reply.topic,
        "phx_reply",
        %{status: reply.status, response: reply.payload}
      ])

    {:socket_push, :binary, data}
  end

  def encode!(%Message{} = msg) do
    data =
      Msgpax.pack!([
        msg.join_ref,
        msg.ref,
        msg.topic,
        msg.event,
        msg.payload
      ])

    {:socket_push, :binary, data}
  end

  @impl true
  def decode!(raw_message, _opts) do
    [join_ref, ref, topic, event, payload | _] = Msgpax.unpack!(raw_message)

    %Message{
      topic: topic,
      event: event,
      payload: payload,
      ref: ref,
      join_ref: join_ref
    }
  end
end
